### ISSUE TITLE
Regression: Proc macros not being expanded

### ISSUE THREAD

**Issue by jkelleyrtp** [rust-lang/rust-analyzer#10894]
It seems that proc macro support degraded heavily from the previous version of RA for me. In the last version, both web-sys autocomplete and proc macro inlay hints worked.

Now, it seems that RA is blind to any web-sys types (once again) *and* none of my proc macros are being expanded (`proc macro not expanded rust-analyzerunresolved-proc-macro`). My proc macro is fairly well behaved and, in the past, had significant IDE support for hovers/go-to/inlay hints. 

For instance, my code 
```rust
let g = rsx!(div {});
```
expands to 
```rust
    let g = NodeFactory::annotate_lazy(move |__cx: NodeFactory| -> VNode {
        __cx.element(dioxus_elements::div, [], [], [], None)
    });
```
I would expect to see hints on `div` since I re-use the token literally, but the macro simply fails to expand.

Is there a different proc-macro server I can use that works?

**rust-analyzer version**: `d9b2291f5 2021-11-29 stable`

**rustc version**: `rustc 1.56.1 (59eed8a2a 2021-11-01)`

<img width="765" alt="Screen Shot 2021-11-30 at 8 34 11 PM" src="https://user-images.githubusercontent.com/10237910/144155617-d26ce506-5714-4f69-9ea7-1a76e66af240.png">

**Comment by jkelleyrtp** [comment]
In general, RA just seems broken across the board for me.

This code should generate an error but doesn't.


<img width="334" alt="Screen Shot 2021-11-30 at 8 44 05 PM" src="https://user-images.githubusercontent.com/10237910/144156509-e0fd96cb-2043-46ca-a30f-4fa656ca74ce.png">

**Comment by Veykril** [comment]
I assume the version of `dioxus` you are using is not the current one on crates io right? As your snippet doesn't compile for me at all.

**Comment by al2me6** [comment]
Are there any updates on this issue? Proc macro expansion seems to have entirely stopped working after I upgraded to `rustc 1.59.0-nightly (928783de6 2021-12-11)`, with both release and nightly RA.

(This can be seen by opening https://github.com/al2me6/evanescence using RA.)

**Comment by lnicola** [comment]
> Are there any updates on this issue?

@al2me6 the proc macro ABI sometimes changes in nightly (which is to be expected), and rust-analyzer needs to be updated.

CC @alexjg

**Comment by lnicola** [comment]
> I'm running an m1 Mac in case RA is downloading the wrong binary.

Are you running the x86 version of Code instead of the native M1 one?

**Comment by lnicola** [comment]
@al2me6 can you try with the next RA nightly? This seems to work now for me.

![image](https://user-images.githubusercontent.com/308347/145852394-816b978a-b5f7-4f62-9cdc-3e4adbf32ce2.png)

@alexjg false alert :-).

**Comment by al2me6** [comment]
That part of the code doesn't have much dependence at all on proc macros.

`evanescence_web/src/state.rs`, for example, seems to be still broken:

![image](https://user-images.githubusercontent.com/35266431/145871285-aa2da855-4b16-4f82-b197-8d432babb1de.png)

With
![image](https://user-images.githubusercontent.com/35266431/145871439-873f43a7-772b-45ef-853b-f83720cd8fd1.png)

**Comment by lnicola** [comment]
@al2me6 you need to try with the _next_ nightly.

**Comment by pwaller** [comment]
FYI I just tried on 791722b70 (verified with version output in vscode) and still don't have proc macros being resolved.

**Comment by bobbbay** [comment]
Bump: I'm getting this a lot and it's making it hard to read my code. Any thoughts on how to fix this?

**Comment by Veykril** [comment]
whats your rust version and rust-analyzer version

**Comment by bobbbay** [comment]
```shell
$ rustc --version
rustc 1.57.0 (f1edd0429 2021-11-29)
$ rust-analyzer --version
rust-analyzer 2021-10-25
```

**Comment by Veykril** [comment]
That rust-analyzer version is heavily outdated which might be the reason for your troubles, try updating to the latest release.

**Comment by bobbbay** [comment]
Updated to `2021-12-27`, the newest version packaged for my distribution. I still have this issue. Should I switch to a manual installation  [via binary](https://rust-analyzer.github.io/manual.html#rust-analyzer-language-server-binary) to get to `2022-01-10`?

**Comment by Veykril** [comment]
That one should probably work. Not sure what could be causing this then without a reproduction/more info

**Comment by vorporeal** [comment]
I'm running into this issue using the nightly toolchain with VSCode rust-analyzer extension.

```sh
$ ~/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/bin/rust-analyzer --version
rust-analyzer 9700add 2022-01-30 dev
```

Example code that reproduces the issue for me:

```rust
use clap::Parser;

#[derive(Parser, Debug)]
struct Args {
}

fn main() {
    let args = Args::parse();
}
```

<img width="508" alt="image" src="https://user-images.githubusercontent.com/211194/151742190-7ffdb091-2d7d-4250-bbac-7579888954d6.png">

**Comment by lukekeum** [comment]
Did you guys tried to download rust-analyzer to your bin and add path to "rust-analyzer.server.path" in "settings.json" which is in vscode ?

This helped me to solve this problem

**Comment by Palladinium** [comment]
I can confirm this is still an issue.

![2022-02-04_12_23_745228021](https://user-images.githubusercontent.com/3919824/152456991-b7c0d815-b08e-4b4d-95bd-3732ece36e74.png)

rust-analyzer was installed with `cargo install rust-analyzer --git https://github.com/rust-analyzer/rust-analyzer.git --locked`
Running in Spacemacs with the Rust layer.
```
$ rustc --version
rustc 1.60.0-nightly (27f5d830e 2022-02-02)
$ rust-analyzer --version
rust-analyzer 9597e55e9 2022-02-04 dev
```
I am positive this is the only version of rust-analyzer I have installed - Spacemacs does not install its own version.

**Comment by flodiebold** [comment]
Works for me with both stable and nightly Rust.

It'd be nice if we could show some more information when we're emitting this error (whether proc macros are disabled would be one thing, but also maybe we're dropping / only logging some errors that can lead to this?)

**Comment by Palladinium** [comment]
Digging a bit deeper it turns out that yes, I simply didn't correctly enable proc-macros in the LSP config.
For anyone else in my same situation, the answer was `(setq-default lsp-rust-analyzer-proc-macro-enable t)`.

I agree that it'd be nice to have more informative errors. Something along the lines of `proc-macro expansion disabled in configuration` would help separate user configuration errors from legitimate rust-analyzer errors.

**Comment by alxiong** [comment]
I also run into this problem on Doom Emacs, I use `rust-analyzer ba3305480 2022-02-07 stable` and `rustc 1.56.1 (59eed8a2a 2021-11-01)`. 

I'm sure my `lsp-rust-analyzer-proc-macro-enable` is set to `t`, in fact I have no problem expand locally declared macro, but failed to correctly expand foreign (imported) proc macro. 

(btw, I also confirm the output of expanded code on vscode via "Rust Analyzer: Expand macro recursively" to perfectly matched that of my emacs, both have no issue with locally defined macro, but both failed with imported macro)

**Comment by flodiebold** [comment]
Let's close this since the original issue is not reproducible, and the ticket risks becoming a dumping ground for proc macro expansion problems.

**Comment by cloutiertyler** [comment]
@flodiebold I think this issue may be related to the M1 Mac aspect of things. I am currently getting these errors in VSCode running the latest version of pre-release rust-analyzer and getting the following error. Seems to be an x86 vs arm issue.
```
[{
	"resource": "./crates/spacetimedb/src/routes/mod.rs",
	"owner": "rustc",
	"code": {
		"value": "unresolved-proc-macro",
		"target": {
			"$mid": 1,
			"external": "https://rust-analyzer.github.io/manual.html#unresolved-proc-macro",
			"path": "/manual.html",
			"scheme": "https",
			"authority": "rust-analyzer.github.io",
			"fragment": "unresolved-proc-macro"
		}
	},
	"severity": 8,
	"message": "proc macro `Deserialize` not expanded: Cannot create expander for ./target/debug/deps/libserde_derive-88182fb8c70c39d8.dylib: Io(Custom { kind: InvalidData, error: DlOpen { desc: \"dlopen(./target/debug/deps/libserde_derive-88182fb8c70c39d8.dylib, 0x000A): tried: \\'./target/debug/deps/libserde_derive-88182fb8c70c39d8.dylib\\' (mach-o file, but is an incompatible architecture (have \\'x86_64\\', need \\'arm64e\\'))\" } })",
	"source": "rust-analyzer",
	"startLineNumber": 19,
	"startColumn": 10,
	"endLineNumber": 19,
	"endColumn": 21
}]```

**Comment by lnicola** [comment]
@cloutiertyler you should file a new issue if you still have trouble. My guess is that you're either running the x86 version of Code, or you have a stale target directory from an x86.

**Comment by maxnoel** [comment]
Hi! as of the latest release, a simple `tokio` "Hello World" program now exhibits the issue:

![image](https://user-images.githubusercontent.com/1538011/174653272-58b2d345-d558-4cc8-bb63-bb764b5537bf.png)

(the program runs fine, the only dependency is `tokio = { version = "1.19.2", features = ["full"] }`)

**Comment by isosphere** [comment]
Chiming in to say I had an issue with this when writing a fresh derive macro, and it went away when I reloaded rust-analyzer (`rust-analyzer: Restart Server` in VSCode). 

My precise output from rust-analyzer was "Proc macro not expanded: no proc macro dynlib present".

**Comment by spraza** [comment]
Just had the same issue and the root cause seemed to be what @joelong01 mentioned above, since removing the stale project fixed the issue for me. +1 to that it'd be great if rust-analyzer points to the right root cause here.

**Comment by shitcodebykaushik** [comment]
The Problem is coming at the root of the project file.
I am still figuring out how to solve with the nighty rust .

**Comment by Veykril** [comment]
Please open a new issue and fill out the template, including what operating system you are on and what error(s) you are getting

**Comment by silvio2402** [comment]
I encountered this issue with `rustc 1.76.0 (07dca489a 2024-02-04)`. After updating to `rustc 1.78.0 (9b00956e5 2024-04-29)` the error is resolved.
